src/Compilers/Core/Portable/PEWriter/MetadataWriter.PortablePdb.cs

```diff
@@ -831,5 +834,100 @@ private void EmbedSourceLink(Stream stream)
+            WriteValue("compilerversion", compilerVersion);
+
+            foreach (var (key, value) in module.CommonCompilation.SerializeForPdb())
+            {
+                WriteValue(key, value);
+            }
+
+            _debugMetadataOpt.AddCustomDebugInformation(
+                parent: EntityHandle.ModuleDefinition,
+                kind: _debugMetadataOpt.GetOrAddGuid(PortableCustomDebugInfoKinds.CompilationOptions),
+                value: _debugMetadataOpt.GetOrAddBlob(builder));
+
+            void WriteValue(string key, string value)
+            {
+                builder.WriteUTF8(key);
+                builder.WriteByte(0);
+                builder.WriteUTF8(value);
+                builder.WriteByte(0);
+            }
+        }
+
+        /// <summary>
+        /// Writes information about metadata references to the pdb so the same
+        /// reference can be found on sourcelink to create the compilation again
+        /// </summary>
+        private void EmbedMetadataReferenceInformation(CommonPEModuleBuilder module)
+        {
+            var builder = new BlobBuilder();
+
+            // Order of information
+            // File name (null terminated string): A.exe
+            // Extern Alias (null terminated string): a1,a2,a3
+            // MetadataImageKind (byte)
+            // EmbedInteropTypes (boolean)
+            // COFF header Timestamp field (4 byte int)
+            // COFF header SizeOfImage field (4 byte int)
+            // MVID (Guid, 24 bytes)
+            foreach (var metadataReference in module.CommonCompilation.ExternalReferences)
+            {
+                if (metadataReference is PortableExecutableReference portableReference && portableReference.FilePath is object)
+                {
+                    var fileName = PathUtilities.GetFileName(portableReference.FilePath);
+                    var reference = module.CommonCompilation.GetAssemblyOrModuleSymbol(portableReference);
+                    var peReader = GetReader(reference);
+
+                    // Don't write before checking that we can get a peReader for the metadata reference
+                    if (peReader is null)
+                        continue;
+
+                    // Write file name first
+                    builder.WriteUTF8(fileName);
+
+                    // Make sure to add null terminator
+                    builder.WriteByte(0);
+
+                    // Extern alias
+                    if (portableReference.Properties.Aliases.Any())
+                        builder.WriteUTF8(string.Join(",", portableReference.Properties.Aliases));
+
+                    // Always null terminate the extern alias list
+                    builder.WriteByte(0);
+
+                    builder.WriteByte((byte)portableReference.Properties.Kind);
+                    builder.WriteBoolean(portableReference.Properties.EmbedInteropTypes);
```

Comment:
Combine these values into a single byte. We only need one bit for each.

Reply:
I don't really think `Kind` will expand much more (if any ever), but it's a byte enum. Should we always assume only two values are possible?

Reply:
We control the type. If we need to add more info that doesn't fit 8 bits we can update the serialization format.

Reply:
Changed to single byte